File: src/game/key.c
/*  Key drivers */

#include <stdio.h>
/*#include <sgtty.h>*/
#include <termios.h>

long save_flag= -1;
char key_buff[256];
char pr_bf[32];
long key_mode= -1;

key_reprint()
{
	extern long pr_due;
	extern long pr_qcr;
	pr_qcr=1;
	pbfr();
	if((key_mode==0)&&(pr_due))
		printf("\n%s%s",pr_bf,key_buff);
	pr_due=0;
	fflush(stdout);
}

File: src/stubs/key.py
"""
Key drivers
"""
# State
global_state = {
    'key_buff': "",
}


need_reprint = False
last_prompt = ''


# Mutations
def set_key_buff(state, key_buff):
    state.update({
        'is_clean': True,
        'key_buff': key_buff,
    })
    return state


# Actions
def key_input(state, prompt, max_length):
    global last_prompt, need_reprint
    last_prompt = prompt
    state = state['bprintf'](state, prompt)
    state = state['pbfr'](state)
    state = set_key_buff(state, input()[:max_length])
    return state


def key_reprint(state):
    state = state['pbfr']({
        **state,
        'pr_qcr': True,
    })
    if state['is_clean']:
        print("\n{}{}".format(last_prompt, state['key_buff']))
    return {
        **state,
        'pr_due': False,
    }
